cadc-vos-server/src/main/java/org/opencadc/vospace/server/pkg/VospacePackageRunner.java

```diff
@@ -0,0 +1,379 @@
+import org.opencadc.vospace.transfer.TransferReader;
+
+public class VospacePackageRunner extends PackageRunner {
+
+    private static final Logger log = Logger.getLogger(VospacePackageRunner.class);
+    private static final String CONTINUING_PROCESSING = ", continuing package processing...";
+    private URI resourceID;
+    private Transfer packageTransfer;
+    private NodePersistence nodePersistence;
+    private VOSpaceAuthorizer vospaceAuthorizer;
+    private PathResolver pathResolver;
+
+    public VospacePackageRunner() {}
+
+    /**
+     * Sets the name of this instance which is used to generate the JNDI key to the NodePersistence,
+     * and initialize other resources used by this class.
+     *
+     * @param appName the name of the instance.
+     */
+    @Override
+    public void setAppName(String appName) {
+        String jndiKey = appName + "-" + NodePersistence.class.getName();
+        try {
+            Context ctx = new InitialContext();
+            this.nodePersistence = (NodePersistence) ctx.lookup(jndiKey);
+            this.vospaceAuthorizer = new VOSpaceAuthorizer(nodePersistence);
+            this.pathResolver = new PathResolver(nodePersistence, vospaceAuthorizer, true);
+            this.resourceID = nodePersistence.getResourceID();
+        } catch (NamingException e) {
+            throw new RuntimeException("BUG: NodePersistence implementation not found with JNDI key " + jndiKey, e);
+        }
+    }
+
+    /**
+     * Get the list of targets for the package and the package name.
+     */
+    @Override
+    protected void initPackage() {
+        log.debug("initPackage start");
+        try {
+            // check job is valid
+            log.debug("job id passed in: " + job.getID());
+            JobInfo jobInfo = job.getJobInfo();
+
+            // Get the target list from the job
+            TransferReader tr = new TransferReader();
+            this.packageTransfer = tr.read(jobInfo.getContent(), VOSURI.SCHEME);
+            List<URI> targetList = packageTransfer.getTargets();
+            if (targetList.size() > 1) {
+                this.packageName = "cadc-download-" + job.getID();
+            } else {
+                this.packageName = getFilenameFromURI(targetList.get(0));
+            }
+            log.debug("package name: " + this.packageName);
+        } catch (IOException | TransferParsingException e) {
+            throw new RuntimeException("ERROR reading jobInfo: ", e);
+        }
+        log.debug("initPackage done");
+    }
+
+    /**
+     * Get the list of PackageItem's to be included in this package.
+     *
+     * @return List of PackageItem's.
+     */
+    @Override
+    protected Iterator<PackageItem> getItems() {
+        // packageTransfer is populated in initPackage
+        List<URI> targetList = packageTransfer.getTargets();
+        log.debug("target list: " + targetList.toString());
+
+        List<PackageItem> packageItems = new ArrayList<>();
```

Comment:
This looks like it accumulates the entire list of items (nodes) in memory before returning. That's going to have high latency (failure mode) and consume an arbitrary amount of memory.

The whole implementation of recursively adding content has to happen inside the Iterator<PackageItem> and traversal has to happen while the PackageRunner iterates.

Reply:
Updated to return an iterator over the PackageItem's.

I left the advance logic in the method. At first used separate methods to refresh the iterators, but it's easier to understand when it's all in one place.